opt/scopeserver/dataserver/utils/loom.py

```diff
@@ -43,37 +43,56 @@ def __init__(self, partial_md5_hash, file_path, abs_file_path, loom_connection,
         # Metrics
         self.nUMI = None
         self.species, self.gene_mappings = self.infer_species()
-        self.ss_pickle_name = os.path.join(os.path.dirname(self.abs_file_path), partial_md5_hash + ".ss_pkl")
+        self.ss_pickle_name = self.abs_file_path.with_suffix(".ss_pkl")
+        self.load_ss(self.ss_pickle_name)
 
+    def load_ss(self, ss_pickle_name: Path) -> None:
         try:
             with open(self.ss_pickle_name, "rb") as fh:
-                logger.debug(f"Loading prebuilt SS for {file_path} from {self.ss_pickle_name}")
+                logger.debug(f"Loading prebuilt SS for {self.file_path} from {self.ss_pickle_name}")
                 self.ss = pickle.load(fh)
-
+            logger.debug(f"Building hsap Search Spaces for {self.file_path}")
+            self.hsap_ss = ss.SearchSpace(loom=self, cross_species="hsap").build()
+            logger.debug(f"Building mmus Search Spaces for {self.file_path}")
+            self.mmus_ss = ss.SearchSpace(loom=self, cross_species="mmus").build()
         except (EOFError, FileNotFoundError):
-            logger.debug(f"Building Search Spaces for {file_path}")
-            if self.species == "dmel":
-                logger.debug(f"Building hsap Search Spaces for {file_path}")
-                self.hsap_ss = ss.SearchSpace(loom=self, cross_species="hsap").build()
-                logger.debug(f"Building mmus Search Spaces for {file_path}")
-
-                self.mmus_ss = ss.SearchSpace(loom=self, cross_species="mmus").build()
-            logger.debug(f"Building self Search Spaces for {file_path}")
-
-            self.ss = ss.SearchSpace(loom=self).build()
-            self.ss.loom = None  # Remove loom connection to enable pickling
-            logger.debug(f"Built all Search Spaces for {file_path}")
-            with open(self.ss_pickle_name, "wb") as fh:
-                logger.debug(f"Writing prebuilt SS for {file_path} to {self.ss_pickle_name}")
-                pickle.dump(self.ss, fh)
+            self.build_all_ss()
+
+    def build_all_ss(self) -> None:
+        logger.debug(f"Building Search Spaces for {self.file_path}")
+        if self.species == "dmel":
+            logger.debug(f"Building hsap Search Spaces for {self.file_path}")
+            self.hsap_ss = ss.SearchSpace(loom=self, cross_species="hsap").build()
+            logger.debug(f"Building mmus Search Spaces for {self.file_path}")
+            self.mmus_ss = ss.SearchSpace(loom=self, cross_species="mmus").build()
+        logger.debug(f"Building self Search Spaces for {self.file_path}")
+
+        self.ss = ss.SearchSpace(loom=self).build()
+        self.write_ss()
+
+    def write_ss(self) -> None:
+        self.ss.loom = None  # Remove loom connection to enable pickling
+        logger.debug(f"Built all Search Spaces for {self.file_path}")
```

Comment:
A copy of the ss should be made here, to prevent an issue with the dictionary being changed during writing. Importantly, the dict.copy() function should NOT be used, as it does not copy the extra attributes we have on the SearchSpace class. The copy module should be used to copy the whole object.